ports/stm/common-hal/sdioio/SDCard.c

```diff
@@ -0,0 +1,356 @@
+            continue;
+        }
+
+        if (reserved_sdio[periph_index-1]) {
+            sdio_taken = true;
+            continue;
+        }
+
+        self->clock = mcu_sdio_clock;
+        self->command = mcu_sdio_command;
+        self->data[0] = mcu_sdio_data0;
+        self->data[1] = mcu_sdio_data1;
+        self->data[2] = mcu_sdio_data2;
+        self->data[3] = mcu_sdio_data3;
+
+        return periph_index;
+    }
+
+    if (sdio_taken) {
+        mp_raise_ValueError(translate("Hardware busy, try alternative pins"));
+    } else {
+        mp_raise_ValueError_varg(translate("Invalid %q pin selection"), MP_QSTR_SDIO);
+    }
+}
+
+
+void common_hal_sdioio_sdcard_construct(sdioio_sdcard_obj_t *self,
+        const mcu_pin_obj_t * clock, const mcu_pin_obj_t * command,
+        uint8_t num_data, mcu_pin_obj_t ** data, uint32_t frequency) {
+
+    int periph_index = check_pins(self, clock, command, num_data, data);
+    SDIO_TypeDef * SDIOx = mcu_sdio_banks[periph_index - 1];
+
+    GPIO_InitTypeDef GPIO_InitStruct = {0};
+
+    // /* GPIOC and GPIOD Periph clock enable */
+    // RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOC | RCC_AHB1Periph_GPIOD | SD_DETECT_GPIO_CLK, ENABLE);
+
+    /* Configure data PC.08, PC.09, PC.10, PC.11 pins: D0, D1, D2, D3 pins */
+    for (int i=0; i<num_data; i++) {
+        GPIO_InitStruct.Pin = pin_mask(data[i]->number);
+        GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+        GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+        GPIO_InitStruct.Pull = GPIO_PULLUP;
+        GPIO_InitStruct.Alternate = self->data[i]->altfn_index;
+        HAL_GPIO_Init(pin_port(data[i]->port), &GPIO_InitStruct);
+    }
+
+    /* Configure PD.02 CMD line */
+    GPIO_InitStruct.Alternate = self->command->altfn_index;
+    GPIO_InitStruct.Pin = pin_mask(command->number);
+    HAL_GPIO_Init(pin_port(command->port), &GPIO_InitStruct);
+
+    /* Configure PC.12 pin: CLK pin */
+    GPIO_InitStruct.Alternate = self->clock->altfn_index;
+    GPIO_InitStruct.Pin = pin_mask(clock->number);
+    HAL_GPIO_Init(pin_port(clock->port), &GPIO_InitStruct);
+
+// XXX hard coded pin
+#define SD_DETECT_PIN GPIO_PIN_12
+#define SD_DETECT_GPIO_PORT GPIOB
+
+    /*!< Configure SD_SPI_DETECT_PIN pin: SD Card detect pin */
+    GPIO_InitStruct.Pin = SD_DETECT_PIN;
+    GPIO_InitStruct.Mode = GPIO_MODE_INPUT;
+    GPIO_InitStruct.Pull = GPIO_PULLUP;
+    HAL_GPIO_Init(SD_DETECT_GPIO_PORT, &GPIO_InitStruct);
+
+    __HAL_RCC_SDIO_CLK_ENABLE();
+
+    self->handle.Init.ClockDiv = SDIO_TRANSFER_CLK_DIV;
+    self->handle.Init.ClockEdge = SDIO_CLOCK_EDGE_RISING;
+    self->handle.Init.ClockBypass = SDIO_CLOCK_BYPASS_DISABLE;
+    self->handle.Init.ClockPowerSave = SDIO_CLOCK_POWER_SAVE_DISABLE;
+    self->handle.Init.BusWide = SDIO_BUS_WIDE_1B;
+    self->handle.Init.HardwareFlowControl = SDIO_HARDWARE_FLOW_CONTROL_DISABLE;
+    self->handle.Instance = SDIOx;
+
+    HAL_StatusTypeDef r = HAL_SD_Init(&self->handle);
+    if (r != HAL_OK) {
```

Comment:
why not just combine these as `if (HAL_SD_Init(&self->handle) != HAL_OK) {...`

Reply:
Because the number stored in r is displayed in the message `"SDIO Init Error %d"`